### PR TITLE
[Feature branch] Support Dataset in `mlflow.evaluate`

### PR DESCRIPTION
**Pull request by prithvikannan**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

* Add `to_evaluation_dataset()` to pyfunc mixin
* Update `mlflow.evaluate` to support an MLflow Dataset input. Log the dataset as input to the run with tag `eval`.
* Update `mlflow.evaluate` to support optional targets if Dataset already defines one.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/evaluation/base.py [hunk 1]
```diff
@@ -1334,12 +1337,18 @@ def pred_sample(eval_df, _builtin_metrics, _artifacts_dir):
...
-        data,
-        targets=targets,
-        path=dataset_path,
-        feature_names=feature_names,
-    )
+    from mlflow.data.pyfunc_dataset_mixin import PyFuncConvertibleDatasetMixin
+
+    if isinstance(data, Dataset) and issubclass(data.__class__, PyFuncConvertibleDatasetMixin):
+        dataset = data.to_evaluation_dataset()
+        mlflow.log_input(data, "eval")
```
Comment (prithvikannan) [inline mlflow/models/evaluation/base.py:1344]: metric prefix

File: mlflow/data/tensorflow_dataset.py [hunk 2]
```diff
@@ -122,6 +122,13 @@ def to_pyfunc(self) -> PyFuncInputsOutputs:
         """
         return PyFuncInputsOutputs(self._data)
 
+    def to_evaluation_dataset(self, path=None, feature_names=None) -> EvaluationDataset:
+        """
+        Converts the dataset to an EvaluationDataset for model evaluation. Required
+        for use with mlflow.sklearn.evalute().
+        """
+        raise NotImplementedError
```
Comment (prithvikannan) [inline mlflow/data/tensorflow_dataset.py:130]: not implemented for now because we need the `targets` field

File: mlflow/data/huggingface_dataset.py [hunk 3]
```diff
@@ -157,6 +158,18 @@ def to_pyfunc(self) -> PyFuncInputsOutputs:
         else:
             return PyFuncInputsOutputs(inputs=df, outputs=None)
 
+    def to_evaluation_dataset(self, path=None, feature_names=None) -> EvaluationDataset:
+        """
+        Converts the dataset to an EvaluationDataset for model evaluation. Required
+        for use with mlflow.sklearn.evalute().
```
Comment (dbczumar) [inline mlflow/data/huggingface_dataset.py:164]:
```suggestion
        for use with mlflow.evalute().
```

Can we apply this fix to the other method docstrings as well?

File: mlflow/models/evaluation/base.py [hunk 4]
```diff
@@ -1334,12 +1337,22 @@ def pred_sample(eval_df, _builtin_metrics, _artifacts_dir):
...
-    )
+    from mlflow.data.pyfunc_dataset_mixin import PyFuncConvertibleDatasetMixin
+
+    if isinstance(data, Dataset) and issubclass(data.__class__, PyFuncConvertibleDatasetMixin):
+        dataset = data.to_evaluation_dataset(dataset_path, feature_names)
+        if evaluator_name_to_conf_map and "default" in evaluator_name_to_conf_map:
+            context = evaluator_name_to_conf_map["default"].get("metric_prefix", None)
+        else:
+            context = None
+        mlflow.log_input(data, context)
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1348]: We're doing this outside of line 1357 (`_start_run_or_reuse_active_run()`), so the run management might not be correct here. Can we move this input logging down into the context created on 1357?

Comment (prithvikannan) [reply mlflow/models/evaluation/base.py]: Done. Also changed to client API

File: tests/models/test_evaluation.py [hunk 5]
```diff
@@ -32,6 +32,9 @@
 
 import mlflow
 from mlflow import MlflowClient
+from mlflow.data.pandas_dataset import from_pandas
+
+# from mlflow.data.pandas_dataset import from_pandas
```
Comment (dbczumar) [inline tests/models/test_evaluation.py:37]: Nit: Remove duplicate  / commented code

```suggestion
```

File: mlflow/models/evaluation/base.py [hunk 6]
```diff
@@ -1102,6 +1106,8 @@ def evaluate(
                    are regarded as feature columns. If it is Spark DataFrame, only the first 10000
                    rows in the Spark DataFrame will be used as evaluation data.
 
+                - A MLflow Dataset object, containing evaluation features and labels.
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1109]: `targets` should be optional if an MLflow Dataset object is specified.

Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1109]:
```suggestion
                - A :py:class`mlflow.data.Dataset` instance containing evaluation features and labels.
```

Comment (dbczumar) [reply mlflow/models/evaluation/base.py:1109]: If an MLflow Dataset is specified but doesn't define `targets`, we should raise a clear error message indicating that either the dataset needs to specify targets or the user needs to pass a value for the targets param of `mlflow.evaluate()`.

**Review by dbczumar — APPROVED** [review]
LGTM once remaining comments are addressed (please ping me before merging). Thanks @prithvikannan !
